### PR TITLE
Make Mlflow client's logging function controlled by global async logging config

### PR DESCRIPTION
**Pull request by chenmoneygithub**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/chenmoneygithub/mlflow/pull/11778?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11778/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11778
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Currently `mlflow.config.enable_async_logging()` only affects fluent API, and it makes sense to have client APIs under control as well. This PR does the work.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [x] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [x] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: mlflow/tracking/client.py [hunk 1]
```diff
@@ -734,7 +735,7 @@ def log_metric(
         value: float,
         timestamp: Optional[int] = None,
         step: Optional[int] = None,
-        synchronous: bool = True,
+        synchronous: Optional[bool] = None,
```
Comment (harupy) [inline mlflow/tracking/client.py:738]: I think the argument description should be updated.

File: tests/tracking/test_client.py [hunk 2]
```diff
@@ -794,3 +795,14 @@ def test_client_log_metric_params_tags_overrides(mock_store):
...
+
+def test_enable_async_logging(mock_store):
+    enable_async_logging(True)
+    MlflowClient().log_param(run_id="run_id", key="key", value="val")
+    mock_store.log_param_async.assert_called_once_with("run_id", Param("key", "val"))
+
+    MlflowClient().log_metric(run_id="run_id", key="key", value="val", step=1, timestamp=1)
+    mock_store.log_metric_async.assert_called_once_with("run_id", Metric("key", "val", 1, 1))
+    # Clean up the setup.
+    enable_async_logging(False)
```
Comment (harupy) [inline tests/tracking/test_client.py:808]: What would happen if we hit an error before `enable_async_logging(False)` is called? Would that affect other tests?

File: mlflow/tracking/client.py (hunk 1 above)
Comment (chenmoneygithub) [reply mlflow/tracking/client.py:738]: it should be fine, the original argument docstring doesn't contain the default value.

Comment (chenmoneygithub) [reply mlflow/tracking/client.py:738]: actually some parts do need a change, let me fix

File: mlflow/tracking/client.py [hunk 3]
```diff
@@ -824,12 +829,13 @@ def log_param(
...
-                successfully. If False, logs the parameter asynchronously and
-                returns a future representing the logging operation.
+            synchronous: *Experimental* If True, blocks until the metric is logged successfully.
+                If False, logs the metric asynchronously and returns a future representing the
+                logging operation. If None, read from environment variable
+                `MLFLOW_ENABLE_ASYNC_LOGGING`, which defaults to False if not set.
 
         Returns:
-            When `synchronous=True`, returns parameter value. When `synchronous=False`,
+            When `synchronous=True` or None, returns parameter value. When `synchronous=False`,
```
Comment (harupy) [inline mlflow/tracking/client.py:838]:
```suggestion
            When `synchronous=True`, returns parameter value. When `synchronous=False`,
```

I think we can remove `or None`

File: tests/tracking/test_client.py [hunk 4]
```diff
@@ -93,6 +94,13 @@ def mock_time():
         yield time
 
 
+@pytest.fixture
+def setup_async_logging():
+    enable_async_logging(True)
+    yield
```
Comment (harupy) [inline tests/tracking/test_client.py:100]:
```suggestion
    yield
    mlflow.flush_async_logging()
```

can we flush here?

**Review by harupy — APPROVED** [review]
Left a few more comments, otherwise LGTM!

**Pull request merged**